zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -816,3 +817,349 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
+                    shift_div_by_64.clone(),
+                    transplacement as u64,
+                    4u64,
+                );
+            for idx in 0..(4 - transplacement) {
+                let tmpidx = idx + transplacement;
+                let merge_a = if idx + transplacement == (3_usize) {
+                    a_slice_front_digits[tmpidx].clone()
+                } else {
+                    a_slice_front_digits[tmpidx].clone()
+                        + a_slice_back_digits[tmpidx + 1].clone()
+                            * shift_mod_by_64_decpow.expr()
+                };
+                shr_constraints[idx] = shr_constraints[idx].clone()
+                    + select_transplacement_polynomial.clone()
+                        * (merge_a - b_digits[idx].clone());
+            }
+            for idx in (4 - transplacement)..4 {
+                shr_constraints[idx] = shr_constraints[idx].clone()
+                    + select_transplacement_polynomial.clone()
+                        * b_digits[idx].clone();
+            }
+        }
+        (0..4).for_each(|idx|
+            cb.require_zero(
+                "merge a_slice_back_digits and a_slice_front_digits == b_digits",
+                shr_constraints[idx].clone(),
+            )
+        );
+
+        // for i in 0..4
+        // a_slice_back_digits[i] + a_slice_front_digits * shift_mod_by_64_pow
+        // == a_digits[i]
+        for idx in 0..4 {
+            cb.require_equal(
+                "a[idx] == a_slice_back[idx] + a_slice_front[idx] * shift_mod_by_64_pow",
+                a_slice_back_digits[idx].clone() + a_slice_front_digits[idx].clone() * shift_mod_by_64_pow.expr(),
+                a_digits[idx].clone(),
+            );
+        }
+
+        // check serveral higher cells == 0 for slice_back and slice_front
+        let mut equal_to_zero = 0.expr();
+        for digit_transplacement in 0..8 {
+            let select_transplacement_polynomial =
+                generate_lagrange_base_polynomial(
+                    shift_mod_by_64_div_by_8.clone(),
+                    digit_transplacement as u64,
+                    8u64,
+                );
+            for virtual_idx in 0..4 {
+                for idx in (digit_transplacement + 1)..8 {
+                    let nowidx = (virtual_idx * 8 + idx) as usize;
+                    equal_to_zero = equal_to_zero
+                        + (select_transplacement_polynomial.clone()
+                            * a_slice_back[nowidx].expr());
+                }
+                for idx in (8 - digit_transplacement)..8 {
+                    let nowidx = (virtual_idx * 8 + idx) as usize;
+                    equal_to_zero = equal_to_zero
+                        + (select_transplacement_polynomial.clone()
+                            * a_slice_front[nowidx].expr());
+                }
+            }
+        }
+        // for i in 1..32
+        // check shift[i] == 0
```

Comment:
This is not correct. The shift can be greater than 256. Just the result will become 0 if shift > 256.